File: src/Web/wwwroot/js/BootstrapHelper.js
/* Copyright (c) Alexandre Kerametlian.
 * Licensed under the Apache License, Version 2.0.
 * 
 * A collection of functions to facilitate interop between Blazor and Bootstrap components.
 */

function BootstrapHelper_ToggleClassOnElement(elementId, classToToggle) {
    var element = document.getElementById(elementId);
    element.classList.toggle(classToToggle);
}

// Maintains a 2-level map: [groupTag -> elementId -> collapse object].
// Keeping the elements groupped by tag lets the caller release the objects 
// in one shot when they're no longer needed.
var collapseMap = {};

function BootstrapHelper_ToggleCollapse(groupTag, elementId) {
    if (collapseMap[groupTag] === undefined) {
        collapseMap[groupTag] = {};
    }

    if (collapseMap[groupTag][elementId] === undefined) {
        var element = document.getElementById(elementId);
        collapseMap[groupTag][elementId] = new bootstrap.Collapse(element);
    }

    collapseMap[groupTag][elementId].toggle();
}

function BootstrapHelper_ReleaseCollapse(groupTag, elementId) {
    if (collapseMap[groupTag] === undefined) {
        return;
    }

    if (collapseMap[groupTag][elementId] !== undefined) {
        collapseMap[groupTag][elementId].dispose();
        delete collapseMap[groupTag][elementId];
    }
}

function BootstrapHelper_ReleaseCollapseGroup(groupTag) {
    if (collapseMap[groupTag] === undefined) {
        return;
    }

    for (const key in collapseMap[groupTag]) {
        collapseMap[groupTag][key].dispose();
    }

    delete collapseMap[groupTag];
}

File: src/Web/BootstrapHelper.cs
// Copyright (c) Alexandre Kerametlian.
// Licensed under the Apache License, Version 2.0.

using Microsoft.JSInterop;

namespace Heimdall.Web;

public class BootstrapHelper
{
    public BootstrapHelper(IJSRuntime jsRuntime)
    {
        this.JSRuntime = jsRuntime;
    }

    private IJSRuntime JSRuntime { get; }

    public async Task ToggleCollapseAsync(string groupTag, string collapseElementId, string toggleButtonElementId = null)
    {
        if (toggleButtonElementId is not null)
        {
            await this.JSRuntime.InvokeVoidAsync(
                "BootstrapHelper_ToggleClassOnElement",
                toggleButtonElementId,
                "collapsed");
        }
        await this.JSRuntime.InvokeVoidAsync("BootstrapHelper_ToggleCollapse", groupTag, collapseElementId);
    }

    public async Task ReleaseCollapseAsync(string groupTag, string elementId)
    {
        await this.JSRuntime.InvokeVoidAsync("BootstrapHelper_ReleaseCollapse", groupTag, elementId);
    }

    public async Task ReleaseCollapseGroupAsync(string groupTag)
    {
        await this.JSRuntime.InvokeVoidAsync("BootstrapHelper_ReleaseCollapseGroup", groupTag);
    }
}
